matching-notification-api/src/main/java/org/phenotips/matchingnotification/internal/DefaultMatchingNotificationManager.java

```diff
@@ -0,0 +1,252 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/
+ */
+package org.phenotips.matchingnotification.internal;
+
+import org.phenotips.data.Patient;
+import org.phenotips.data.PatientRepository;
+import org.phenotips.data.permissions.PermissionsManager;
+import org.phenotips.data.permissions.Visibility;
+import org.phenotips.data.similarity.PatientSimilarityView;
+import org.phenotips.matchingnotification.MatchingNotificationManager;
+import org.phenotips.matchingnotification.finder.MatchFinderManager;
+import org.phenotips.matchingnotification.match.PatientMatch;
+import org.phenotips.matchingnotification.match.internal.DefaultPatientMatch;
+import org.phenotips.matchingnotification.notification.PatientMatchEmail;
+import org.phenotips.matchingnotification.notification.PatientMatchNotificationResponse;
+import org.phenotips.matchingnotification.notification.PatientMatchNotifier;
+import org.phenotips.matchingnotification.storage.MatchStorageManager;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.query.Query;
+import org.xwiki.query.QueryException;
+import org.xwiki.query.QueryManager;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.hibernate.HibernateException;
+import org.hibernate.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Joiner;
+
+/**
+ * @version $Id$
+ */
+@Component
+@Singleton
+public class DefaultMatchingNotificationManager implements MatchingNotificationManager
+{
+    private Logger logger = LoggerFactory.getLogger(DefaultMatchingNotificationManager.class);
+
+    @Inject
+    private QueryManager qm;
+
+    @Inject
+    private PatientRepository patientRepository;
+
+    @Inject
+    private PermissionsManager permissionsManager;
+
+    @Inject
+    @Named("matchable")
+    private Visibility matchableVisibility;
+
+    @Inject
+    private PatientMatchNotifier notifier;
+
+    @Inject
+    private MatchFinderManager matchFinderManager;
+
+    @Inject
+    private MatchStorageManager matchStorageManager;
+
+    @Override
+    public List<PatientMatch> findAndSaveMatches(double score)
+    {
+        List<PatientMatch> addedMatches = new LinkedList<>();
+
+        List<Patient> patients = this.getPatientsList();
+        for (Patient patient : patients) {
```

Comment:
Added a check. I think the patient would not be added to the list because of the visibility condition, but it should be more obvious in the code.